### DISCUSSION TITLE
How to override the text-color of an el-button for type primary (using scss variable)?

### DISCUSSION THREAD

**Discussion by Skuallpa** [element-plus/element-plus#11628] [Q&A] [answered]
I'm trying to override the text-color of the `el-button` by overriding the appropriate scss variable. 

```scss
$--button: (
  'text-color': red
);

@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $button: $--button,

);

@use 'element-plus/theme-chalk/src/index.scss' as *;
```

This works when the button as no "type" (primary, danger, ...), but when applying the `primary` type, the text color is not applied (defaults to white). How can I override the text-color for a primary button?

```html
<el-button type="primary">Primary</el-button>
```

Thanks in advance for your help

**Comment by btea** [comment, marked as answer]
Perhaps you can override the default by customizing the css variable.[example](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtcm93IGNsYXNzPVwibWItNFwiPlxuICAgIDxlbC1idXR0b24gdHlwZT1cInByaW1hcnlcIj5QcmltYXJ5PC9lbC1idXR0b24+XG4gIDwvZWwtcm93PlxuPC90ZW1wbGF0ZT5cbjxzdHlsZT5cbiAgLmVsLWJ1dHRvbi5lbC1idXR0b24tLXByaW1hcnkge1xuICAgIC0tZWwtYnV0dG9uLXRleHQtY29sb3I6IGFxdWE7XG4gIH1cbjwvc3R5bGU+IiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJfbyI6e319)

**Comment by Skuallpa** [comment]
Thanks for your answer. Yes, that would be a solution, but I'd like to override scss variables as much as possible and avoid having to override css variables (like `.el-button.el-button--primary`).

Isn't it possible to override some of the [element plus scss variables](https://github.com/element-plus/element-plus-vite-starter/blob/main/src/styles/element/index.scss) to achieve this behaviour?

Isn't it possible to do something similar as for the colors definition? I mean, for the colors, we can do:



```scss
$--colors: (
  'primary': (
    'base': green,
  ),
);


@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $colors: $--colors,

);

@use 'element-plus/theme-chalk/src/index.scss' as *;

```

I've tried doing something similar for the button:

```scss
$--button: (
  'primary': (
    'text-color': green
  )
);

@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $button: $--button,

);

@use 'element-plus/theme-chalk/src/index.scss' as *;
````

But that doesn't work. I get an error that ("text-color": green) isn't a valid CSS value.

Any idea?

**Comment by ryuhangyeong** [comment]
I don't know exactly your environment, but I've confirmed that you can redefine the theme normally in the environment below.

## development environment

- `create vue`
- `sass`(^1.58.2)

### Code

- `main.ts`

```ts
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'
import ElementPlus from 'element-plus'

import './index.scss'

const app = createApp(App)

app.use(router)
app.use(ElementPlus)
app.mount('#app')
```

- `index.scss`

```scss
@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $main: (
    'padding': 100px,
  ),
  $button: (
    'hover-bg-color': red,
  )
);

@use 'element-plus/theme-chalk/src/index.scss' as *;
```

- `App.vue`

```vue
<template>
  <el-button>el-button</el-button>
  <el-main>el-main</el-main>
</template>
```

<img width="844" alt="image" src="https://user-images.githubusercontent.com/27342882/219636231-b124703e-533f-4469-a82e-1096abf745f9.png">

Is this what you want?

**Comment by Skuallpa** [comment]
Yes, that works also on my side as well.

My question is how to change the text color of a button of **type primary** using the scss variables exposed by element-plus?

I mean for such a button:
```html
<el-button type="primary">Primary</el-button>
```
As described in my first comment, I am able to change the text color of a "non-primary" button using:

```scss
$--button: (
  'text-color': red
);

@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $button: $--button,

);

@use 'element-plus/theme-chalk/src/index.scss' as *;
```
But that doesn't apply on a button of type "primary".

Thanks in advance for your help.

**Comment by btea** [comment]
This doesn't seem to be possible, and there is no relevant variable for the button text configuration. The currently existing text variable configurations are only the following: 
https://github.com/element-plus/element-plus/blob/aba0d26537447b936775554e5afe5d9b6d53e33c/packages/theme-chalk/src/common/var.scss#L96-L106
This affects other component texts if you make related configuration overrides.Therefore, I think you're better off doing exactly what I suggested at the beginning, with custom css variables for exact overrides.

**Comment by Skuallpa** [comment]
Thanks for your answer.

Looking at the code, it seems it was intended (or foresee) to be be possible to customize the `text-color` of the button for the different types (see `$button-text-color` at line 674). However, the variable is not handled in the loop over the types (see line 675)

https://github.com/element-plus/element-plus/blob/19e5bb842569f42513f0496b321ed83a6f043c74/packages/theme-chalk/src/common/var.scss#L642-L689

I will then override the css variable directly as you suggested.

```css
  .el-button.el-button--primary {
    --el-button-text-color: aqua;
  }
```

Thanks for your help.

**Comment by btea** [comment]
That seems to be the case, so I marked the original reply as answered.
